File: src/ui/controllers/MainWindowController.java
package ui.controllers;

import board.Board;
import board.Board.Site;
import board.BoardFactory;
import game.Game;
import game.ui.GameUI;
import javafx.fxml.FXML;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.TilePane;
import javafx.scene.layout.VBox;
import javafx.stage.Modality;
import javafx.stage.Stage;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import players.Player;
import players.PlayerUI;
import ui.components.Cell;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collection;

public class MainWindowController implements GameUI, PlayerUI {
    @FXML
    private AnchorPane anchorPane;
    @FXML
    private TilePane tilePane;
    @FXML
    protected Label whiteScore;
    @FXML
    protected Label blackScore;
    @FXML
    protected Label currentPlayerLabel;

    private static final int BOARD_SIZE = 8;
    private Player black;
    private Player white;
    private Game game;
    private ArrayList<Cell> cells;

    public void setGame(Game game) {
        this.game = game;
    }
    public void setPlayers(final Player black, final Player white) {
        this.black = black;
        this.white = white;
    }

    private Site getCurrentSite(Player currentPlayer) {
        return currentPlayer.equals(black) ? Site.PLAYER : Site.OPPONENT;
    }

    private void addCells() {
        cells = new ArrayList<Cell>();
        for (int i = 0; i < BOARD_SIZE; i++)
            for (int j = 0; j < BOARD_SIZE; j++) {
                final Cell cell;
                cell = new Cell(i, j);
                cell.showDisk();
                cells.add(cell);
                tilePane.getChildren().addAll(cell);
            }
    }

    @FXML
    public void initialize() {
        tilePane.getChildren().clear();
        tilePane.setPadding(new Insets(20, 20, 20, 20));
        tilePane.setPrefColumns(8);
        tilePane.setPrefRows(8);
        tilePane.setMaxSize(380, 380);
        tilePane.setStyle("-fx-background-color: forestgreen;");

        final BoardFactory boardFactory = new BoardFactory();
        Board board = boardFactory.getStartingBoard();

        addCells();
        game.evaluate(board,black);
    }

    @Override
    public void endGame(@Nullable Player winner) {
        Stage primaryStage = (Stage) anchorPane.getScene().getWindow();
        final Stage dialog = new Stage();
        dialog.initModality(Modality.APPLICATION_MODAL);
        dialog.initOwner(primaryStage);
        dialog.setResizable(false);
        VBox dialogVbox = new VBox(20);

        String gameEndMessage;
        if (winner == null) gameEndMessage = "Draw!";
        else gameEndMessage = "Player " + winner.getName() + " wins!";

        Label message = new Label(gameEndMessage);
        message.setAlignment(Pos.CENTER);
        dialogVbox.getChildren().add(message);
        dialog.setScene(new Scene(dialogVbox, 260, 40));
        dialog.show();
    }

    @Override
    public void printCurrentPlayer(@NotNull Player player) {
        currentPlayerLabel.setText(player.getName());
    }

    @Override
    public void displayBoard(@NotNull Board board, @NotNull Player currentPlayer) {
        for (int i = 0; i < BOARD_SIZE; i++)
            for (int j = 0; j < BOARD_SIZE; j++) {
                final Site site = board.getSite(i * BOARD_SIZE + j);
                final Cell cell = cells.get(i*BOARD_SIZE + j);
                if (site != null) {
                    final String color = site.equals(getCurrentSite(currentPlayer)) ? "black" : "white";
                    cell.setColor(color);
                    cell.showDisk();
                } else {
                    cell.hideDisk();
                }
            }

        blackScore.setText(String.valueOf(board.getScore(Site.PLAYER)));
        whiteScore.setText(String.valueOf(board.getScore(Site.OPPONENT)));
    }

    @NotNull
    @Override
    public Player nextPlayer(@NotNull Player currentPlayer) {
        return currentPlayer.equals(black) ? white : black;
    }

    @Override
    public void beginGame(@NotNull Player black, @NotNull Player white) {}

    @Override
    public void representMove(int move, @NotNull Function0<Unit> onDecided) {
        Cell cell = cells.get(move);
        cell.setColor("grey");
        cell.setOnClick(onDecided);
        cell.showDisk();
    }
}

File: src/ui/controllers/MainWindowController.kt
package ui.controllers

import board.Board
import game.Game
import game.ui.GameUI
import javafx.fxml.FXML
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.Label
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.TilePane
import javafx.scene.layout.VBox
import javafx.stage.Modality
import javafx.stage.Stage
import players.Player
import players.PlayerUI
import ui.components.Cell
import java.util.*

class MainWindowController:GameUI,PlayerUI {


	@FXML private lateinit var anchorPane:AnchorPane
	@FXML private lateinit var tilePane:TilePane
	@FXML private lateinit var whiteScore:Label
	@FXML private lateinit var blackScore:Label
	@FXML private lateinit var currentPlayerLabel:Label
	private lateinit var black:Player
	private lateinit var white:Player
	private lateinit var game:Game
	private val cells=ArrayList<Cell>()

	fun setGame(game:Game) {
		this.game=game
	}

	private fun initialiseCells() {
		cells.clear()
		for(i in 0 until Board.BOARD_SIZE)
			for(j in 0 until Board.BOARD_SIZE) {
				val cell=Cell()
				cell.setDiskVisible(false)
				cells.add(cell)
				tilePane.children.addAll(cell)
			}
	}

	@FXML
	fun initialize() {
		tilePane.children.clear()
		tilePane.padding=Insets(20.0,20.0,20.0,20.0)
		tilePane.prefColumns=8
		tilePane.prefRows=8
		tilePane.setMaxSize(380.0,380.0)
		tilePane.style="-fx-background-color: forestgreen;"
		initialiseCells()
		game.playGame()
	}

	override fun endGame(winner:Player?) {
		val primaryStage=anchorPane.scene.window as Stage
		val dialog=Stage()
		dialog.initModality(Modality.APPLICATION_MODAL)
		dialog.initOwner(primaryStage)
		dialog.isResizable=false
		val dialogVBox=VBox(20.0)

		val gameEndMessage=if(winner==null)
			"Draw!"
		else
			"Player "+winner.name+" wins!"

		val message=Label(gameEndMessage)
		message.alignment=Pos.CENTER
		dialogVBox.children.add(message)
		dialog.scene=Scene(dialogVBox,260.0,40.0)
		dialog.show()
	}

	override fun printCurrentPlayer(player:Player) {
		currentPlayerLabel.text=player.name
	}

	override fun displayBoard(board:Board,currentPlayer:Player) {

		val blackSite=if(currentPlayer==black) Board.Site.PLAYER else Board.Site.OPPONENT
		val whiteSite=if(currentPlayer==white) Board.Site.PLAYER else Board.Site.OPPONENT

		for(i in 0 until Board.BOARD_SIZE)
			for(j in 0 until Board.BOARD_SIZE) {
				val site=board.getSite(i*Board.BOARD_SIZE+j)
				val cell=cells[i*Board.BOARD_SIZE+j]
				if(site!=null) {
					val color=if(site==blackSite)
						"black"
					else
						"white"
					cell.setColor(color)
				}
				cell.setDiskVisible(site!=null)
				cell.setOnClick(null)
				cell.setText("")
			}



		blackScore.text=board.getScore(blackSite).toString()
		whiteScore.text=board.getScore(whiteSite).toString()
	}

	override fun markLatestPlayedMove(move:Int) {
		val cell=cells[move]
		cell.setText("I")
	}

	override fun beginGame(black:Player,white:Player) {
		this.black=black
		this.white=white
	}

	override fun representMove(move:Int,onDecided:() -> Unit) {
		val cell=cells[move]
		cell.setColor("grey")
		cell.setOnClick(onDecided)
		cell.setDiskVisible(true)
	}
}
